src/frontend/src/instance.rs

```diff
@@ -331,6 +348,112 @@ impl Instance {
             .await
             .context(TableOperationSnafu)
     }
+
+    async fn check_otlp_legacy(
+        &self,
+        names: &Vec<&String>,
+        ctx: QueryContextRef,
+    ) -> server_error::Result<bool> {
+        let db_string = ctx.get_db_string();
+        let cache = self
+            .otlp_metrics_table_legacy_cache
+            .entry(db_string)
+            .or_default();
+
+        // check cache
+        let hit_cache = names
+            .iter()
+            .filter_map(|name| cache.get(*name))
+            .collect::<Vec<_>>();
+        if !hit_cache.is_empty() {
+            let hit_legacy = hit_cache.iter().any(|en| *en.value());
+            let hit_prom = hit_cache.iter().any(|en| !*en.value());
+
+            // hit but have true and false, means both legacy and new mode are used
+            // we cannot handle this case, so return error
+            // add doc links in err msg later
+            ensure!(!(hit_legacy && hit_prom), OtlpMetricModeIncompatibleSnafu);
+
+            let flag = hit_legacy;
+            // set cache for all names
+            names.iter().for_each(|name| {
+                if !cache.contains_key(*name) {
+                    cache.insert(name.to_string(), flag);
+                }
+            });
+            return Ok(flag);
+        }
```

Comment:
Is it possible to cache whether the database has legacy tables only?
If there is a cache for the database, use the cached result, otherwise we check the table info for result and fill the cache.

Reply:
This can still cause inconsistency. For example: 
1. the agent collects metric A, B, C from source 1 using the old OTLP metric path.
2. the cluster gets upgraded and restarted, all caches are gone
3. the agent collects metric E, F from source 2, since they're new, use the new compatible path

Now we have a database with both format of data. The next time A, B and C come in, it will error because the table schema is wrong.